robot_calibration/src/freight_ground_finder.cpp

```diff
@@ -0,0 +1,180 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <robot_calibration/capture/freight_ground_finder.h>
+#include <sensor_msgs/point_cloud2_iterator.h>
+#include <math.h>
+
+namespace robot_calibration
+{
+
+const unsigned X = 0;
+const unsigned Y = 1;
+const unsigned Z = 2;
+
+FreightGroundFinder::FreightGroundFinder(ros::NodeHandle & nh) :
+  FeatureFinder(nh),
+  waiting_(false)
+{
+  std::string topic_name;
+  nh.param<std::string>("topic", topic_name, "/points");
+  subscriber_ = nh.subscribe(topic_name,
+                             1,
+                             &FreightGroundFinder::cameraCallback,
+                             this);
+
+  nh.param<std::string>("camera_sensor_name", camera_sensor_name_, "camera");
+  nh.param<std::string>("chain_sensor_name", chain_sensor_name_, "ground");
+
+  publisher_ = nh.advertise<sensor_msgs::PointCloud2>("ground_plane_points", 10);
+  if (!depth_camera_manager_.init(nh))
+  {
+    // Error will be printed in manager
+    throw;
+  }
+}
+
+void FreightGroundFinder::cameraCallback(const sensor_msgs::PointCloud2& cloud)
+{
+  if (waiting_)
+  {
+    cloud_ = cloud;
+    waiting_ = false;
+  }
+}
+
+bool FreightGroundFinder::waitForCloud()
+{
+  ros::Duration(1/10.0).sleep();
+
+  waiting_ = true;
+  int count = 250;
+  while (--count)
+  {
+    if (!waiting_)
+    {
+      // success
+      return true;
+    }
+    ros::Duration(0.01).sleep();
+    ros::spinOnce();
+  }
+  ROS_ERROR("Failed to get cloud");
+  return !waiting_;
+}
+
+
+bool FreightGroundFinder::find(robot_calibration_msgs::CalibrationData * msg)
+{
+  geometry_msgs::PointStamped rgbd;
+  geometry_msgs::PointStamped world;
+
+  if (!waitForCloud())
+  {
+    ROS_ERROR("No point cloud data");
+    return false;
+  }
+
+  //  Remove NaNs in the point cloud
+  size_t num_points = cloud_.width * cloud_.height;
+  sensor_msgs::PointCloud2ConstIterator<float> xyz(cloud_, "x");
+  sensor_msgs::PointCloud2Iterator<float> iter(cloud_, "x");
+
+  size_t j = 0;
+  for (size_t i = 0; i < num_points; i++)
+  {
+    geometry_msgs::Point p;
+    p.x = (xyz + i)[X];
+    p.y = (xyz + i)[Y];
+    p.z = (xyz + i)[Z];
+
+    if (!std::isfinite(p.x) || !std::isfinite(p.y) || !std::isfinite(p.z))
+      continue;
+
+    if (p.z > 1.8)  //  pmd doesnt see the ground at a further distance
+      continue;
+    (iter + j)[X] = (xyz + i)[X];
+    (iter + j)[Y] = (xyz + i)[Y];
+    (iter + j)[Z] = (xyz + i)[Z];
+    j++;
+  }
+
+  cloud_.height = 1;
+  cloud_.width  = j;
+  cloud_.data.resize(cloud_.width *  cloud_.point_step);
+
+  int points_total = j;
```

Comment:
looks like there are only two changes in this finder compared to the original ground plane -- this line, and the "if (p.z > 1.8)" above -- if that is the case, we should add two parameters to the original ground plane finder --
 
 * "max_points" - so that this line becomes "int points_total = std::min(j, max_points)
 * "max_z_range" - so that if max_z_range > 0, we always drop points with p.z > max_z_range (and then we default max_z_range to -1 to preserve backwards compatibility)